File: clojure/robot-simulator/src/robot_simulator.clj
(ns robot-simulator)

(defn robot [coordinates, bearing] 
  {
    :coordinates coordinates
    :bearing bearing
    }
)

(defn turn-right [direction]
  (cond 
   (= :north direction) :east
   (= :east direction) :south
   (= :south direction) :west
   (= :west direction) :north
  )
)

(defn turn-left [direction]
  (cond 
   (= :north direction) :west
   (= :west direction) :south
   (= :south direction) :east
   (= :east direction) :north
  )
)

(defn advance [robbie]
  (def bearing (:bearing robbie))
  (def x (:x (:coordinates robbie)))
  (def y (:y (:coordinates robbie)))

  (cond
    (= :north bearing) (robot {:x x :y (+ y 1)} bearing)
    (= :east bearing) (robot {:x (+ x 1) :y y} bearing)
    (= :south bearing) (robot {:x x :y (- y 1)} bearing)
    (= :west bearing) (robot {:x (- x 1) :y y} bearing)
  )     
        
)

(defn apply-move [move, robbie] 
  (cond
    (= "L" move) (robot (:coordinates robbie) (turn-left (:bearing robbie)))
    (= "R" move) (robot (:coordinates robbie) (turn-right (:bearing robbie)))
    (= "A" move) (advance robbie)
  )
)

(defn do-simulate [moves_list, robbie]
  (if (empty? moves_list) 
    robbie
    (do-simulate (rest moves_list) (apply-move (first moves_list) robbie))
  )
)

(defn simulate [moves, robbie] 
  (def moves_list (clojure.string/split moves #""))
  (do-simulate moves_list robbie)
)

File: python/robot-simulator/robot_simulator.py
# Globals for the directions
# Change the values as you see fit
NORTH = 0
EAST = 1
SOUTH = 2
WEST = 3

class Robot:
    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self.direction = direction
        self.coordinates = (x_pos, y_pos)

    def move(self, moves):
        for m in moves:
            self.exec_move(m)

    def exec_move(self, direction):
        match direction:
            case "R":
                self.direction = (self.direction + 1) % 4

            case "L":
                self.direction = (self.direction + 3) % 4

            case "A":
                self.advance()

    def advance(self):
        d = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        x, y = self.coordinates
        dx, dy = d[self.direction]
        self.coordinates = (x + dx, y + dy)
